graph_djikstra_heap.c

```diff
@@ -0,0 +1,240 @@
+        perror("Heap is full\n");
+        abort();
+    }
+
+    if((self->size > 0) && (self->size % HEAP_INIT_SIZE == 0)) {
+        heap_realloc(self);
+    }
+
+    self->p_queue[self->size] = priority_key_arr;
+    self->size++;
+
+    if(self->size > 1) {
+        heap_heapify(self);
+    }
+}
+
+void heap_heapify(Heap *self) {
+    int heap_size = self->size;
+    int** heap_values = self->p_queue;
+
+    int i, index, left, right, smallest;
+
+    // Start from the last non-leaf node to compare it to it's child(ren) and move upwards
+    for (i = (heap_size - 1) / 2; i >= 0; i--) {
+        index = i;
+        // printf("parent index distance val = %d\n", heap_values[index][0]);
+
+        while (1) {
+            left = 2 * index + 1;
+            right = 2 * index + 2;
+            //suppose that current element is the smallest between parent node value and left/right child
+            smallest = index;
+
+            // Compare with left child
+            if (left < heap_size && heap_values[left][0] < heap_values[smallest][0]) {
+                smallest = left;
+            }
+
+            // Compare with right child: if smaller than left child than replace the left child
+            if (right < heap_size && heap_values[right][0] < heap_values[smallest][0]) {
+                smallest = right;
+            }
+
+            // printf("smallest distance val = %d\n", heap_values[smallest][0]);
+
+            //smallest is parent: at this iteration step heap property is restored
+            if (smallest == index) {
+                break;
+            }
+            //the smallest is not the parent, swap index pointer and smallest
+            int *tmp = heap_values[index];
+            heap_values[index] = heap_values[smallest];
+            heap_values[smallest] = tmp;
+
+            index = smallest;  // Move down: there may be elements smaller than swapped value
+        }
+    }
+}
+
+Heap* heap_create() {
+    Heap *heap = (Heap*) malloc(sizeof(Heap));
+    if(heap == NULL) {
+        perror("Failed to allocate memory for heap");
+        abort();
+    }
+
+    //Allocate memory in heap for array keeps 10 pointers to int arrays
+    int **p_queue = malloc(HEAP_INIT_SIZE * sizeof(int *));
+    if(p_queue == NULL) {
+        perror("Failed to allocate memory for heap external array");
+        abort();
+    }
+
+    // // Allocate memory for each inner array FIXME
+    // for (int i = 0; i < HEAP_INIT_SIZE; i++) {
+    //     p_queue[i] = (int *) malloc(P_QUEUE_ELEMENT_SIZE * sizeof(int));
+
+    //     if (p_queue[i] == NULL) {
+    //         perror("Failed to allocate memory for heap inner array");
+    //         abort();
+    //     }
+    // }
+
+    heap->p_queue = p_queue;
+    heap->size = 0;
+
+    return heap;
+}
+
+int main() {
+    Heap *heap = heap_create();
+
+    //emulate djikstra's interaction with heap
+    //start vertice
+    int* start = (int *) malloc(P_QUEUE_ELEMENT_SIZE * sizeof(int)); //TODO make macro
```

Comment:
you dont need malloc here. int start[2] would be fine. But why? ask yourself and you gonna be wise